python/plugins/processing/algs/qgis/TilesXYZ.py

```diff
@@ -0,0 +1,366 @@
+
+        wgs_crs = QgsCoordinateReferenceSystem('EPSG:4326')
+        dest_crs = QgsCoordinateReferenceSystem('EPSG:3857')
+
+        src_to_wgs = QgsCoordinateTransform(project.crs(), wgs_crs, context.transformContext())
+        wgs_to_dest = QgsCoordinateTransform(wgs_crs, dest_crs, context.transformContext())
+
+        settings = QgsMapSettings()
+        settings.setOutputImageFormat(QImage.Format_ARGB32_Premultiplied)
+        settings.setDestinationCrs(dest_crs)
+        settings.setLayers(layers)
+        settings.setOutputDpi(dpi)
+
+        wgs_extent = src_to_wgs.transform(extent)
+        wgs_extent = [wgs_extent.xMinimum(), wgs_extent.yMinimum(), wgs_extent.xMaximum(), wgs_extent.yMaximum()]
+
+        metatiles_by_zoom = {}
+        metatiles_count = 0
+        for zoom in range(min_zoom, max_zoom + 1):
+            metatiles = get_metatiles(wgs_extent, zoom, 4)
+            metatiles_by_zoom[zoom] = metatiles
+            metatiles_count += len(metatiles)
+
+        lab_buffer_px = 100
+        progress = 0
+
+        tile_params = {
+            'format': tile_format,
+            'quality': 75,
+            'width': tile_width,
+            'height': tile_height
+        }
+        if output_format == 'Directory':
+            writer = DirectoryWriter(os.path.join(output_dir, name), tile_params)
+        else:
+            filename = os.path.join(output_dir, '%s.mbtiles' % name)
+            writer = MBTilesWriter(filename, tile_params, wgs_extent, min_zoom, max_zoom)
+
+        for zoom in range(min_zoom, max_zoom + 1):
+            feedback.pushConsoleInfo('Generating tiles for zoom level: %s' % zoom)
+
+            for i, metatile in enumerate(metatiles_by_zoom[zoom]):
+                size = QSize(tile_width * metatile.rows(), tile_height * metatile.columns())
+                extent = QgsRectangle(*metatile.extent())
+                # TODO: transformation of bounding points?
+                settings.setExtent(wgs_to_dest.transform(extent))
+                settings.setOutputSize(size)
+
+                label_area = QgsRectangle(settings.extent())
+                lab_buffer = label_area.width() * (lab_buffer_px / size.width())
+                label_area.set(
+                    label_area.xMinimum() + lab_buffer,
+                    label_area.yMinimum() + lab_buffer,
+                    label_area.xMaximum() - lab_buffer,
+                    label_area.yMaximum() - lab_buffer
+                )
+                settings.setLabelBoundaryGeometry(QgsGeometry.fromRect(label_area))
+
+                image = QImage(size, QImage.Format_ARGB32_Premultiplied)
+                image.fill(Qt.transparent)
+                dpm = settings.outputDpi() / 25.4 * 1000
+                image.setDotsPerMeterX(dpm)
+                image.setDotsPerMeterY(dpm)
+                painter = QPainter(image)
+                job = QgsMapRendererCustomPainterJob(settings, painter)
```

Comment:
I don't think this is safe to do from a background thread -- I think you need to setup ALL the jobs in advance, in the prepareAlgorithm step

Reply:
@nyalldawson this would mean allocating all tiles at different zoom level in advance! for a mobile project ti may work, but not rendering a big area. I've experience with Barcelona municipality and it generate many GBs from 10 to 21 zoom level. Any clue to prepare jobs without this memory allocatio? something like a job queue prepared in the main thread and the processor/renderer as consumer?

Reply:
Yeah I was thinking about the same - right now we don't have a way to modify map settings of an existing map renderer job and creating great amount of these in advance is probably going to be very wasteful of resources. Looking at the Rasterize algorithm, it also creates QgsMapRendererCustomPainterJob for each tile it renders within processAlgorithm.

While I fully agree this is suboptimal and potentially unsafe, it works fine and it's probably better to keep it this way until there's API to handle this...